### ISSUE TITLE
[1.7.10] Permission to fork and use in a public modpack?

### ISSUE THREAD

**Issue by cosmicdan** [Funwayguy/Epic-Siege-Mod#47]
Hi,

Love ESM, but I wanted to code in a function to only give mobs the "Siege" capabilities during a full moon. Not sure of the license you have though.

If I added this config, would you accept a PR for it? If you can't be bothered even touching the old 1.7.10 branch anymore, can I do it myself and just use it exclusively in the modpack we're building (and permission to distribute the unofficial fork with the final modpack package via a custom launcher)?

Either way, we would assume 100% liability for any bugs and problems - of course it will remain as an open fork here on GitHub and retain credits.

**Comment by Funwayguy** [comment]
I'm somewhat open to the idea but do you have a plan on how to implement such a feature? My old 1.7.10 code is far from clean nor very flexible

**Comment by cosmicdan** [comment]
Just a check in the shouldExecute() method for the siege-related AI tasks would be good enough, I'd think? First check a static bool in config to see if it's enabled, if so then check the moon phase in the world provider.

**Comment by Dulciphi** [comment]
Your code is simply the best we've found for the purpose, Funwayguy. 
To be specific, it is the dig and pillar AI we'd wish to limit to full moons. This would give players more time to rebuild and prepare, and also ties in well with the pack's lore.

**Comment by cosmicdan** [comment]
Main considerations are this:

1) Should the "is it a full moon? if so, let's add siege capability" check be performed at the spawn level (i.e. when you add the siege-related AI tasks), or within the AI tasks themselves? The former is more performative, but the latter should allow already-spawned mobs to gain the siege capability once the fullmoon actually happens, but I'm not sure this would be necessary or even desirable.
2) What AI tasks should actually be *considered* when determining what siege AI is only possible in the full moon? It seems Zombies are the only creature capable of siege AI, so I'm looking at the changes it makes here - https://github.com/Funwayguy/Epic-Siege-Mod/blob/1.7.10/src/main/java/funwayguy/esm/core/ESM_Utils.java#L316 - which would also be the same spot I'd check for fullmoon if it'll be checked on-spawn rather than in-task.

Although, one thing Funwayguy. @Dulciphi tells me that the config allows you to whitelist any entity to give them siege capability, e.g. Skeletons can be set to Dig. But I did a recursive search on the entire source tree and the only case where ESM_EntityAIDigging is used is on that block that adds tasks to EntityZombies. I can't actually find any logic that applied siege AI tasks to whitelisted mobs. Am I blind or is your code really that dirty :p

**Comment by cosmicdan** [comment]
Sorry, edited a lot - deleted the old comment. Didn't want a reply to the old comment.

Main considerations are this:

1) Should the "is it a full moon? if so, let's add siege capability" check be performed at the spawn level (i.e. when you add the siege-related AI tasks), or within the AI tasks themselves? The former is more performative, but the latter should allow already-spawned mobs to gain the siege capability once the fullmoon actually happens - although I'm not sure this would be necessary or even desirable.
2) What AI tasks should actually be *considered* when determining what siege AI is only possible in the full moon? It seems Zombies are the only creature capable of siege AI, so I'm looking at the changes it makes here - https://github.com/Funwayguy/Epic-Siege-Mod/blob/1.7.10/src/main/java/funwayguy/esm/core/ESM_Utils.java#L316 - which would also be the same spot I'd check for fullmoon if it'll be checked on-spawn rather than in-task.

I suppose #2 is ideal. I am worried about performance (potentially a hundred calls per-tick to check if it's a fullmoon) but I forget how good JIT optimization can be sometimes... I guess that would that be OK.

Although, one thing Funwayguy. @Dulciphi tells me that the config allows you to whitelist any entity to give them siege capability, e.g. Skeletons can be set to Dig. But I did a recursive search on the entire source tree and the only case where ESM_EntityAIDigging is used is on that block that adds tasks to EntityZombies. I can't actually find any logic that applied siege AI tasks to whitelisted mobs, even searched for all instances of calling addTask. Am I blind or is your code really that dirty :p

I was going to start work on it, but that is a blocker.

**Comment by Funwayguy** [comment]
You might be looking at the unfinished 1.10.2 code which is very clean in comparison and does have capabilities for other mobs to digs (including Creepers which I tested).

I'd go with option 2 but instead use some math on the world time (because full moons are always on a multiple of 8)

You can make your own fork but remove the update notifications completely and make sure you change the title of the mod to reflect that it's a modified version. I'm not fond of having to redirect bug reports because of identical names

**Comment by cosmicdan** [comment]
Fair enough, she must have been mistaken then (there are a few other mods like yours).

It would be better to use world#getMoonPhase. It internally does a modulo operation on the world time anyway (returns an int from 0-7), it's better performance-wise to re-use an existing method - JIT caches return values, and modulo is a relatively expensive operator :)

Thanks a lot, I appreciate it - we might actually make other changes in future too so that's fantastic. I can still do a PR if you want this moonphase option in the official though.

**Comment by cosmicdan** [comment]
Alright, I've made the change I was after. I realized that a "on full moon" option was pretty inflexible, so opted for the math-on-worldtime approach (which is probably want you meant in the first place). Added a every x days config and a y days warmup config.

https://github.com/HostileNetworks/Epic-Siege-Mod_ReignModpack/commit/d588d1791324a2f221ed7ba2aa9a6c8864a84341

If it looks good to you I'll send a PR.
